### PR TITLE
SSD Accuracy Fix

### PR DESCRIPTION
**Pull request by swetha097**
This fixes the SSD training's accuracy issue

### CODE REVIEW

File: rocAL/rocAL/source/randombboxcrop_reader.cpp [hunk 1]
```diff
@@ -111,15 +112,16 @@ pCropCord RandomBBoxCropReader::get_crop_cord(const std::string &image_name)
     return it->second;
 }
 
-void RandomBBoxCropReader::add(std::string image_name, BoundingBoxCord crop_box)
+int RandomBBoxCropReader::add(std::string image_name, BoundingBoxCord crop_box)
```
Comment (rrawther) [inline rocAL/rocAL/source/randombboxcrop_reader.cpp:115]: This implementation can't update the crop is there is already a crop exists. We need to change the crop box update when loading images.

File: rocAL/rocAL_pybind/amd/rali/plugin/pytorch.py [hunk 2]
```diff
@@ -22,6 +22,7 @@ def next(self):
 
     def __next__(self):
         if b.getRemainingImages(self.loader._handle) < self.bs:
+            b.raliResetLoaders(self.loader._handle)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rali/plugin/pytorch.py:25]: this looks like a hack to fix the issue. The raliResetLoaders() has to be called in a seperate function which can be called at the end of every epoch. I think the Pytorch training code has already this routine and we can use similar code

Comment (rrawther) [reply rocAL/rocAL_pybind/amd/rali/plugin/pytorch.py:25]: Please call the reset() function instead

File: rocAL/rocAL_pybind/amd/rali/plugin/tf.py [hunk 3]
```diff
@@ -19,6 +19,7 @@ def next(self):
 
     def __next__(self):
         if b.getRemainingImages(self.loader._handle) < self.bs:
+            b.raliResetLoaders(self.loader._handle)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rali/plugin/tf.py:22]: remove this call from here and call reset() instead

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 4]
```diff
@@ -276,8 +277,8 @@ def draw_patches(img,idx, bboxes):
         thickness = 2
         image = cv2.UMat(image).get()
         image = cv2.rectangle(image, (int(loc_[0]*wtot ),int( loc_[1] *htot)),(int((loc_[2] *wtot) ) ,int((loc_[3] *htot) )) , color, thickness)  
-        #cv2.imwrite(str(idx)+"_"+"train"+".png", 255*image)
-        cv2.imwrite(str(idx)+"_"+"train"+".png", image)
+        cv2.imwrite(str(idx)+"_"+"train"+".png", 255*image)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:280]: this is not a correct inverse normalization value for the image. The output was washed out when displayed.

**Review by rrawther — CHANGES_REQUESTED** [review]
@swetha097 : I suggest making the changes I have suggested in the meeting for random bbox reader. Though it fixes the accuracy, it is too slow.

File: rocAL/rocAL/source/cifar10_data_loader.cpp [hunk 5]
```diff
@@ -303,3 +303,8 @@ decoded_image_info CIFAR10DataLoader::get_decode_image_info()
     return _output_decoded_img_info;
 }
 
+
+crop_image_info CIFAR10DataLoader::get_crop_image_info()
+{
+    return _circ_buff.get_cropped_image_info();
+}
```
Comment (rrawther) [inline rocAL/rocAL/source/cifar10_data_loader.cpp:310]: add newline here

File: rocAL/rocAL/source/image_loader.cpp [hunk 6]
```diff
@@ -312,3 +322,8 @@ decoded_image_info ImageLoader::get_decode_image_info()
 {
     return _output_decoded_img_info;
 }
+
+crop_image_info ImageLoader::get_crop_image_info()
+{
+    return _output_cropped_img_info;
+}
```
Comment (rrawther) [inline rocAL/rocAL/source/image_loader.cpp:329]: add newline here

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 7]
```diff
@@ -175,7 +175,8 @@ def __next__(self):
         self.img_size = np.zeros((self.bs * 2), dtype="int32")
         self.loader.GetImgSizes(self.img_size)
         print("Image sizes:", self.img_size)
-        #img = torch.from_numpy(self.out)
+        # Draw images: Make sure to uncomment the below line to dump images
+        img = torch.from_numpy(self.out)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:179]: Can you please update the changes from my branch for this. Basically we need to make it optional based on a flag and do the normalization correct to view output properly.

**Review by rrawther — CHANGES_REQUESTED** [review]
Please incorporate the requested changes

File: rocAL/rocAL/source/cifar10_data_loader.cpp (hunk 5 above)
Comment (swetha097) [reply rocAL/rocAL/source/cifar10_data_loader.cpp]: Done

File: rocAL/rocAL/source/image_loader.cpp (hunk 6 above)
Comment (swetha097) [reply rocAL/rocAL/source/image_loader.cpp]: Done

File: rocAL/rocAL_pybind/example/coco_pipeline.py (hunk 7 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/example/coco_pipeline.py]: Done

**Review by rrawther — APPROVED** [review]
@kiritigowda : Please merge this PR once @paveltc verify the unit tests for rocal.
I will fix the static analysis warning in the next PR

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 8]
```diff
@@ -9,14 +9,12 @@
...
 import amd.rali.ops as ops
 import amd.rali.types as types
-
-
 import sys
 import numpy as np
 
 
 class COCOPipeline(Pipeline):
-    def __init__(self, batch_size, num_threads, device_id, seed, data_dir, ann_dir, default_boxes,  crop, rali_cpu=True):
```
Comment (kiritigowda) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:19]: @swetha097 can you look at this line - https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=7580035

**Comment by rrawther** [comment]
@swetha097: Can you please merge the tot changes to this PR?

**Comment by swetha097** [comment]
> @swetha097: Can you please merge the tot changes to this PR?

Done

**Review by rrawther — APPROVED** [review]

**Review by kiritigowda — APPROVED** [review]
